grblas/ops.py

```diff
@@ -137,31 +200,29 @@ class OpBase:
 
     def __init__(self, name):
         self.name = name
-        self._specific_types = {}
+        self._typed_ops = {}
+        self.types = {}
 
     def __repr__(self):
         return f'{type(self).__name__}.{self.name}'
 
     def __getitem__(self, type_):
-        type_ = self._normalize_type(type_)
-        if type_ not in self._specific_types:
+        type_ = _normalize_type(type_)
+        if type_ not in self._typed_ops:
             raise KeyError(f'{self.name} does not work with {type_}')
-        return self._specific_types[type_]
+        return self._typed_ops[type_]
 
-    def __setitem__(self, type_, obj):
-        type_ = self._normalize_type(type_)
-        self._specific_types[type_] = obj
+    def _add(self, op):
+        self._typed_ops[op.type] = op
+        self.types[op.type] = op.return_type
 
     def __delitem__(self, type_):
-        type_ = self._normalize_type(type_)
-        del self._specific_types[type_]
+        type_ = _normalize_type(type_)
+        del self._typed_ops[type_]
```

Comment:
Should `type_` also be removed from `self.types` here?

Reply:
Good catch!